proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+
+Input/Output Shaper (aka Shaper) is a new pipeline component that can invoke arbitrary, registered functions, on the
+invocation context (query, documents etc.) of a pipeline and pass the new/modified variables further down the pipeline.
+
+# Basic example
+
+In the following example, we use InputOutputShaper to add a new variable `questions` to the invocation context.
+`questions` is a copy of query variable. This functionality of InputOutputShaper is useful when we simply want to
+rename a variable in the invocation context e.g. in cases where the PromptNode template is expecting a variable
+'questions' rather than 'query'.
+
+
+```python
+
+    from haystack import Pipeline, Document
+
+    with open("tmp_config.yml", "w") as tmp_file:
+        tmp_file.write(
+            f"""
+            version: ignore
+            components:
+            - name: shaper
+              params:
+                inputs:
+                    query:
+                      output: questions
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+        """
+        )
+    pipeline = Pipeline.load_from_yaml(path="tmp_config.yml")
+    result = pipeline.run(
+        query="What can you tell me about Berlin?",
+        documents=[Document("Berlin is an amazing city."), Document("I love Berlin.")],
+    )
+    assert result
+    # query has been renamed to questions
+    assert isinstance(result["meta"]["invocation_context"]["questions"], str)
+
+
+```
+
+# Motivation
+
+We need InputOutputShaper to support the use cases where we want to easily add new variables to the invocation context.
+These new variables hold values which are a result of some arbitrary function invocation on the existing variables
+in the invocation context.
+
+InputOutputShaper is especially useful when combined with PromptNodes. Aside from simply renaming variables to match
+the templates of PromptNodes, we can also use InputOutputShaper to add new variables to the invocation context. Often
+these new variables are the result of some arbitrary function invocation on the existing variables in the invocation.
+
+The original idea for InputOutputShaper was to support the use case of expanding the query string variable to a list of
+strings and the list size matching the size of the documents list. This is useful when we want to use the same query string
+as a question to pose to all the documents in the documents list.
+
+The expected outcome of using InputOutputShaper is that we can easily add new variables to the invocation context
+so they can match the prompt templates of PromptNodes. Multiple InputOutputShaper components can be used in a pipeline to
+modify the invocation context as needed.
+
+
+# Detailed design
+
+The InputOutputShaper component is primarily used in pipelines and configured via a YAML file. The YAML component
```

Comment:
Why "primarily"? Where else would it be used?

Reply:
I meant to say in pipeline definitions; perhaps the wording is off.  But you are right, noted!